rust/agents/scraper/src/chain_scraper/mod.rs

```diff
@@ -150,20 +149,23 @@ impl HyperlaneSqlDb {
         let mut txns_to_fetch = txns.iter_mut().filter(|(_, id)| id.0.is_none());
 
         let mut txns_to_insert: Vec<StorableTxn> = Vec::with_capacity(CHUNK_SIZE);
+        let mut hashes_to_insert: Vec<&H256> = Vec::with_capacity(CHUNK_SIZE);
+
         for mut chunk in as_chunks::<(&H256, &mut (Option<i64>, i64))>(txns_to_fetch, CHUNK_SIZE) {
             for (hash, (_, block_id)) in chunk.iter() {
                 let info = self.provider.get_txn_by_hash(hash).await?;
+                hashes_to_insert.push(*hash);
                 txns_to_insert.push(StorableTxn {
                     info,
                     block_id: *block_id,
                 });
             }
 
-            let mut cur_id = self.db.store_txns(txns_to_insert.drain(..)).await?;
-            for (_hash, (txn_id, _block_id)) in chunk.iter_mut() {
-                debug_assert!(cur_id > 0);
-                let _ = txn_id.insert(cur_id);
-                cur_id += 1;
+            self.db.store_txns(txns_to_insert.drain(..)).await?;
+            let ids = self.db.get_txn_ids(hashes_to_insert.drain(..)).await?;
+
+            for (hash, (txn_id, _block_id)) in chunk.iter_mut() {
+                let _ = txn_id.insert(ids[hash]);
```

Comment:
So we used to make the assumption that `Insert::many` in `sea_orm` works atomically and the inserted txn_ids are consecutive, when in fact they're not because of concurrent updates?

Reply:
Insert many does work atomically, the issue isn't that we have two simultaneous inserts but that we have two inserts with overlapping data. We used to make sure that we would not query the data if it was in the DB, but now we can't fully do that because

1) A checks database to see if h1, h2, and h3 are present
2) A finds out they are not
3) A begins querying for h1, h2, and h3 from the provider
4) B checks the database to see if h1, h4 are present
5) B finds out they are not
6) B begins querying for h1, h4 from the provider
7) A finishes querying for h1, h2, and h3 and inserts them
8) B finishes querying for h1 and h4 and fails to insert them because h1 is not a hash conflict

The issue is not that 7 and 8 happen together, it is because (1, 2, 3, 7) overlaps with (4, 5, 6, 8).